vm/devices/support/fs/lxutil/src/windows/path.rs

```diff
@@ -74,6 +78,59 @@ fn char_needs_escape(c: u8) -> bool {
     index < NTFS_LEGAL_ANSI_CHARACTERS.len() && !NTFS_LEGAL_ANSI_CHARACTERS[index]
 }
 
+// Check if a character needs to be unescaped.
+fn char_needs_unescape(c: u16) -> bool {
+    // Not all characters within the range correspond to escaped characters,
+    // so  make sure that the original character needed to be escaped.
+    return c >= PATH_ESCAPE_MIN
+        && c <= PATH_ESCAPE_MAX
+        && char_needs_escape((c - PATH_ESCAPE_MIN) as u8);
+}
+
+// Convert the path seperators from NT to LX.
+pub fn nt_path_to_lx_path(path: &UnicodeString) -> lx::Result<UnicodeString> {
+    // Copy the path into a new UnicodeString
+    let mut new_path = UnicodeString::new(path.as_slice()).map_err(|_| lx::Error::EINVAL)?;
+    let path_slice = new_path.as_mut_slice();
+    for c in path_slice {
+        if *c == '\\' as u16 {
+            *c = '/' as u16;
+        }
+    }
+
+    Ok(new_path)
+}
+
+// Unescape a path. This function is only used in the lxutil implementation,
+// so the parameter and return types are `UnicodeString`s rather than `Path`s.
+// This also makes the conversion a lot simpler, as the conversions use a
+// single u16 value to represent the escape codepoint.
+//
+// If the path doesn't need unescaping, return None.
+pub fn unescape_path(path: &UnicodeString) -> lx::Result<Option<UnicodeString>> {
```

Comment:
Why not just operate directly on a `&mut UnicodeString` or, better yet, a `&mut [u16]`? Let the caller manage the allocation if a fresh one is needed.

Reply:
Alright, I think this was bad guidance, sorry Will. Now, combined with my other feedback to use `UnicodeStringRef` in a bunch of places, both of the callers are allocating just to have a mutable buffer for this.

I think we need to add some better `UNICODE_STRING` wrappers to make this easier to write. Ideally you'd be able to say something like `fn unescape_path(path: &UnicodeStr) -> Cow<'_, UnicodeString>`, but that's not possible with our current set of types.

I'm OK with this as-is for this change.

Reply:
Actually, if I look at the callers here, they're both immediately turning the result into a `String`. So maybe `unescape_path` should just always do that conversion, something like:

```
pub fn unescape_path(path: &[u16]) -> Result<String> {
    char::decode_utf16(path.iter().map(|c| if char_needs_unescape(c) { c - PATH_ESCAPE_MIN } else { c })).collect()
}
```

Reply:
(Again, perhaps skipping unescaping the / like lxutil.dll)